File: python/src/data_platform_lab/streaming/processor.py
"""Streaming-style sensor event processor.

Reads JSONL sensor events, validates, deduplicates, routes rejected/duplicate
events to a dead-letter file, computes per-sensor aggregates, and writes a
run summary.
"""

from __future__ import annotations

import json
import logging
import time
from collections.abc import Iterable, Iterator
from dataclasses import asdict, dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from dataexcept import FileReadError, FileWriteError

from data_platform_lab.manifest import write_manifest

logger = logging.getLogger(__name__)

REQUIRED_FIELDS: list[str] = [
    "sensor_id",
    "type",
    "value",
    "unit",
    "location",
    "timestamp",
]


@dataclass
class EventResult:
    """Result for a single processed event."""

    event: dict[str, Any]
    status: str
    reason: str | None = None


@dataclass
class StreamSummary:
    """Run summary — follows platform conventions (snake_case fields)."""

    pipeline_name: str = "sensor_stream"
    run_at: str = ""
    duration_seconds: float = 0.0
    status: str = "success"
    events_seen: int = 0
    events_accepted: int = 0
    events_rejected: int = 0
    events_duplicate: int = 0
    dead_letter_count: int = 0
    events_late: int = 0
    max_lateness_seconds: float = 0.0
    watermark: str = ""
    lateness_threshold_seconds: float = 0.0
    aggregates: dict[str, Any] = field(default_factory=dict)
    rejection_reasons: dict[str, int] = field(default_factory=dict)
    manifest_path: str = ""


def parse_event_time(timestamp_str: str) -> datetime:
    """Parse an ISO 8601 timestamp string into a timezone-aware datetime."""
    return datetime.fromisoformat(timestamp_str.replace("Z", "+00:00"))


def classify_lateness(
    event_time: datetime,
    watermark: datetime | None,
    threshold_seconds: float,
) -> tuple[bool, float]:
    """Determine if an event is late relative to the current watermark."""
    if watermark is None:
        return False, 0.0
    lateness = (watermark - event_time).total_seconds()
    is_late = lateness > threshold_seconds
    return is_late, max(lateness, 0.0)


def validate_event(event: dict[str, Any]) -> EventResult:
    """Validate a single event against required-field and type rules."""
    for fld in REQUIRED_FIELDS:
        if fld not in event:
            return EventResult(event=event, status="rejected", reason=f"missing field: {fld}")

    if event["value"] is None:
        return EventResult(event=event, status="rejected", reason="null value")

    for fld in REQUIRED_FIELDS:
        if fld == "value":
            continue
        if not isinstance(event[fld], str) or not event[fld].strip():
            return EventResult(
                event=event, status="rejected", reason=f"empty or invalid field: {fld}"
            )

    if not isinstance(event["value"], (int, float)):
        return EventResult(event=event, status="rejected", reason="value is not a number")

    try:
        datetime.fromisoformat(event["timestamp"].replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return EventResult(event=event, status="rejected", reason="unparseable timestamp")

    return EventResult(event=event, status="accepted")


def deduplicate_key(event: dict[str, Any]) -> str:
    """Return a dedup key: ``"{sensor_id}::{timestamp}"``."""
    return f"{event['sensor_id']}::{event['timestamp']}"


def compute_aggregates(events: list[dict[str, Any]]) -> dict[str, Any]:
    """Compute grouped aggregates over accepted events."""
    by_sensor: dict[str, dict[str, Any]] = {}
    by_type: dict[str, int] = {}
    by_location: dict[str, int] = {}

    for evt in events:
        sid = evt["sensor_id"]
        val = evt["value"]
        if sid not in by_sensor:
            by_sensor[sid] = {"count": 0, "min_value": val, "max_value": val, "_sum": 0.0}
        entry = by_sensor[sid]
        entry["count"] += 1
        entry["_sum"] += val
        if val < entry["min_value"]:
            entry["min_value"] = val
        if val > entry["max_value"]:
            entry["max_value"] = val

        event_type = evt["type"]
        by_type[event_type] = by_type.get(event_type, 0) + 1
        location = evt["location"]
        by_location[location] = by_location.get(location, 0) + 1

    for entry in by_sensor.values():
        entry["avg_value"] = round(entry["_sum"] / entry["count"], 2)
        del entry["_sum"]

    return {"by_sensor": by_sensor, "by_type": by_type, "by_location": by_location}


def _read_lines(path: Path) -> Iterator[str]:
    """Yield source lines while preserving file-level failure context."""
    try:
        with path.open(encoding="utf-8") as fh:
            yield from fh
    except OSError as exc:
        raise FileReadError(str(path), exc) from exc


def _ensure_output_dir(path: Path) -> None:
    try:
        path.mkdir(parents=True, exist_ok=True)
    except OSError as exc:
        raise FileWriteError(str(path), exc) from exc


def _write_lines(path: Path, lines: Iterable[str]) -> None:
    """Write text chunks incrementally while preserving file failure context."""
    try:
        with path.open("w", encoding="utf-8") as fh:
            for line in lines:
                fh.write(line)
    except OSError as exc:
        raise FileWriteError(str(path), exc) from exc


def _write_text(path: Path, text: str) -> None:
    _write_lines(path, (text,))


def process_stream(
    input_path: str | Path,
    output_dir: str | Path,
    pipeline_name: str = "sensor_stream",
    lateness_threshold_seconds: float = 0.0,
) -> StreamSummary:
    """Process a JSONL file of sensor events end-to-end."""
    input_path = Path(input_path)
    output_dir = Path(output_dir)
    _ensure_output_dir(output_dir)

    logger.info("Starting pipeline '%s' — reading from %s", pipeline_name, input_path)
    start = time.monotonic()

    results: list[EventResult] = []
    seen_keys: set[str] = set()
    accepted_events: list[dict[str, Any]] = []
    rejection_reasons: dict[str, int] = {}
    watermark: datetime | None = None
    late_events: list[dict[str, Any]] = []
    max_lateness = 0.0

    for raw_line in _read_lines(input_path):
        line = raw_line.strip()
        if not line:
            continue

        try:
            event = json.loads(line)
        except json.JSONDecodeError:
            result = EventResult(event={"_raw": line}, status="rejected", reason="malformed JSON")
            rejection_reasons["malformed JSON"] = rejection_reasons.get("malformed JSON", 0) + 1
            logger.warning("Rejected event (malformed JSON): %s", line[:120])
            results.append(result)
            continue

        try:
            result = validate_event(event)
        except Exception:
            result = EventResult(event=event, status="rejected", reason="validation error")
            logger.warning("Rejected event (validation error): %s", event)

        if result.status == "rejected":
            reason = result.reason or "unknown"
            rejection_reasons[reason] = rejection_reasons.get(reason, 0) + 1
            logger.warning("Rejected event (%s): %s", reason, event)
            results.append(result)
            continue

        key = deduplicate_key(event)
        if key in seen_keys:
            result = EventResult(event=event, status="duplicate", reason="duplicate event")
            rejection_reasons["duplicate event"] = rejection_reasons.get("duplicate event", 0) + 1
            logger.warning("Duplicate event: %s", key)
            results.append(result)
            continue

        seen_keys.add(key)
        accepted_events.append(event)
        event_time = parse_event_time(event["timestamp"])
        is_late, lateness = classify_lateness(event_time, watermark, lateness_threshold_seconds)
        if is_late:
            late_events.append(event)
            if lateness > max_lateness:
                max_lateness = lateness
        if watermark is None or event_time > watermark:
            watermark = event_time
        results.append(result)

    aggregates = compute_aggregates(accepted_events)
    accepted_path = output_dir / "accepted.jsonl"
    dead_letter_path = output_dir / "dead_letter.jsonl"
    late_events_path = output_dir / "late_events.jsonl"

    _write_lines(accepted_path, (json.dumps(evt) + "\n" for evt in accepted_events))
    _write_lines(
        dead_letter_path,
        (
            json.dumps({"event": result.event, "status": result.status, "reason": result.reason})
            + "\n"
            for result in results
            if result.status in ("rejected", "duplicate")
        ),
    )
    _write_lines(late_events_path, (json.dumps(evt) + "\n" for evt in late_events))

    duration = time.monotonic() - start
    events_accepted = sum(1 for result in results if result.status == "accepted")
    events_rejected = sum(1 for result in results if result.status == "rejected")
    events_duplicate = sum(1 for result in results if result.status == "duplicate")

    summary = StreamSummary(
        pipeline_name=pipeline_name,
        run_at=datetime.now(UTC).isoformat(),
        duration_seconds=round(duration, 4),
        status="success",
        events_seen=len(results),
        events_accepted=events_accepted,
        events_rejected=events_rejected,
        events_duplicate=events_duplicate,
        dead_letter_count=events_rejected + events_duplicate,
        events_late=len(late_events),
        max_lateness_seconds=round(max_lateness, 2),
        watermark=watermark.isoformat() if watermark else "",
        lateness_threshold_seconds=lateness_threshold_seconds,
        aggregates=aggregates,
        rejection_reasons=rejection_reasons,
    )

    summary_path = output_dir / "summary.json"
    _write_text(summary_path, json.dumps(asdict(summary), indent=2))

    logger.info(
        "Pipeline '%s' complete — %d accepted (%d late), %d rejected, %d duplicate (%.3fs)",
        pipeline_name,
        events_accepted,
        len(late_events),
        events_rejected,
        events_duplicate,
        duration,
    )

    manifest_path = write_manifest(
        pipeline_name=pipeline_name,
        run_id=datetime.now(UTC).strftime("%Y%m%d_%H%M%S"),
        source=str(input_path),
        output=str(accepted_path),
        row_count=events_accepted,
        status="success",
        extras={
            "events_seen": len(results),
            "events_rejected": events_rejected,
            "events_duplicate": events_duplicate,
            "events_late": len(late_events),
            "dead_letter_path": str(dead_letter_path),
        },
    )
    summary.manifest_path = str(manifest_path)
    return summary

File: javascript/src/streaming/processor.js
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { join } from "node:path";

/**
 * @typedef {object} SensorEvent
 * @property {string} sensor_id
 * @property {string} type
 * @property {*}      value
 * @property {string} unit
 * @property {string} location
 * @property {string} timestamp
 */

/**
 * @typedef {object} ValidationResult
 * @property {SensorEvent} event
 * @property {"accepted"|"rejected"|"duplicate"} status
 * @property {string|null} reason
 */

/**
 * @typedef {object} SensorAggregates
 * @property {Object<string, {count: number, min_value: number, max_value: number, avg_value: number}>} by_sensor
 * @property {Object<string, number>} by_type
 * @property {Object<string, number>} by_location
 */

/**
 * @typedef {object} PipelineSummary
 * @property {string}  pipeline_name
 * @property {string}  run_at
 * @property {number}  duration_seconds
 * @property {string}  status
 * @property {number}  events_seen
 * @property {number}  events_accepted
 * @property {number}  events_rejected
 * @property {number}  events_duplicate
 * @property {number}  dead_letter_count
 * @property {number}  events_late
 * @property {number}  max_lateness_seconds
 * @property {string}  watermark
 * @property {number}  lateness_threshold_seconds
 * @property {SensorAggregates} aggregates
 * @property {Object<string, number>} rejection_reasons
 */

/**
 * Check whether a timestamp string is parseable as a valid date.
 *
 * @param {string} ts - Timestamp string to validate.
 * @returns {boolean} True if the timestamp is valid.
 */
function isValidTimestamp(ts) {
  try {
    const d = new Date(ts);
    const iso = d.toISOString();
    return iso !== "Invalid Date";
  } catch {
    return false;
  }
}

/**
 * Parse an ISO 8601 timestamp string into a Date object.
 *
 * @param {string} ts - ISO timestamp string.
 * @returns {Date} Parsed Date object.
 */
export function parseEventTime(ts) {
  return new Date(ts);
}

/**
 * Determine if an event is late relative to the current watermark.
 *
 * @param {Date} eventTime - The event's timestamp as a Date.
 * @param {Date|null} watermark - The current watermark (max event time seen).
 * @param {number} thresholdSeconds - Allowed lateness in seconds.
 * @returns {{is_late: boolean, lateness_seconds: number}}
 */
export function classifyLateness(eventTime, watermark, thresholdSeconds) {
  if (watermark === null) {
    return { is_late: false, lateness_seconds: 0 };
  }
  const latenessMs = watermark.getTime() - eventTime.getTime();
  const latenessSeconds = Math.max(latenessMs / 1000, 0);
  const isLate = latenessSeconds > thresholdSeconds;
  return { is_late: isLate, lateness_seconds: latenessSeconds };
}

/**
 * Validate a single sensor event against the required-field schema.
 *
 * Checks that all required fields are present and well-typed:
 * - sensor_id, type, unit, location, timestamp must be non-empty strings
 * - value must be a finite number (not null, undefined, NaN, or Infinity)
 * - timestamp must be parseable as an ISO date
 *
 * @param {SensorEvent} event - The event object to validate.
 * @returns {ValidationResult} The event with an accepted/rejected status.
 */
export function validateEvent(event) {
  // Check string fields
  const stringFields = ["sensor_id", "type", "unit", "location", "timestamp"];
  for (const field of stringFields) {
    const val = event[field];
    if (typeof val !== "string" || val.trim() === "") {
      return {
        event,
        status: "rejected",
        reason: `missing or empty field: ${field}`,
      };
    }
  }

  // Check value is a number (not null, undefined, NaN, etc.)
  if (event.value === null || event.value === undefined) {
    return { event, status: "rejected", reason: "null value" };
  }
  if (typeof event.value !== "number" || !Number.isFinite(event.value)) {
    return { event, status: "rejected", reason: "value is not a valid number" };
  }

  // Check timestamp is parseable
  if (!isValidTimestamp(event.timestamp)) {
    return { event, status: "rejected", reason: "unparseable timestamp" };
  }

  return { event, status: "accepted", reason: null };
}

/**
 * Compute a deduplication key for a sensor event.
 *
 * The key is formed as `"{sensor_id}::{timestamp}"`, which identifies
 * logically identical readings from the same sensor at the same instant.
 *
 * @param {SensorEvent} event - The event to derive a key from.
 * @returns {string} A composite deduplication key.
 */
export function deduplicateKey(event) {
  return `${event.sensor_id}::${event.timestamp}`;
}

/**
 * Compute aggregate statistics over a set of accepted sensor events.
 *
 * Returns breakdowns by sensor (with min/max/avg value), by event type
 * (count), and by location (count).
 *
 * @param {SensorEvent[]} events - Array of accepted events.
 * @returns {SensorAggregates} Aggregated statistics.
 */
export function computeAggregates(events) {
  /** @type {Object<string, {count: number, sum: number, min_value: number, max_value: number}>} */
  const bySensor = {};
  /** @type {Object<string, number>} */
  const byType = {};
  /** @type {Object<string, number>} */
  const byLocation = {};

  for (const evt of events) {
    // by_sensor
    if (!bySensor[evt.sensor_id]) {
      bySensor[evt.sensor_id] = {
        count: 0,
        sum: 0,
        min_value: evt.value,
        max_value: evt.value,
      };
    }
    const s = bySensor[evt.sensor_id];
    s.count += 1;
    s.sum += evt.value;
    if (evt.value < s.min_value) s.min_value = evt.value;
    if (evt.value > s.max_value) s.max_value = evt.value;

    // by_type
    byType[evt.type] = (byType[evt.type] || 0) + 1;

    // by_location
    byLocation[evt.location] = (byLocation[evt.location] || 0) + 1;
  }

  // Build the final by_sensor with avg_value rounded to 2 decimals
  /** @type {Object<string, {count: number, min_value: number, max_value: number, avg_value: number}>} */
  const bySensorFinal = {};
  for (const [sensorId, data] of Object.entries(bySensor)) {
    bySensorFinal[sensorId] = {
      count: data.count,
      min_value: data.min_value,
      max_value: data.max_value,
      avg_value: Math.round((data.sum / data.count) * 100) / 100,
    };
  }

  return {
    by_sensor: bySensorFinal,
    by_type: byType,
    by_location: byLocation,
  };
}

/**
 * Process a JSONL stream of sensor events end-to-end.
 *
 * Reads events line-by-line from the input file, validates them, deduplicates
 * (first occurrence wins), routes accepted events to `accepted.jsonl` and
 * rejected/duplicate events to `dead_letter.jsonl`, computes aggregates over
 * accepted events, and writes a summary to `summary.json`.
 *
 * @param {string} inputPath - Path to the JSONL input file.
 * @param {string} outputDir - Directory to write output files into.
 * @param {object} [options]
 * @param {string} [options.pipelineName="sensor_stream"] - Name for the pipeline run.
 * @returns {Promise<PipelineSummary>} The pipeline run summary.
 */
export async function processStream(inputPath, outputDir, { pipelineName = "sensor_stream", latenessThresholdSeconds = 0 } = {}) {
  const startTime = performance.now();
  const runAt = new Date().toISOString();

  console.info(`[${pipelineName}] Starting stream processing from ${inputPath}`);

  // Ensure output directory exists
  await mkdir(outputDir, { recursive: true });

  /** @type {SensorEvent[]} */
  const acceptedEvents = [];
  /** @type {Array<{event: *, status: string, reason: string}>} */
  const deadLetterEvents = [];
  /** @type {Object<string, number>} */
  const rejectionReasons = {};
  const seenKeys = new Set();

  let eventsSeen = 0;
  let eventsAccepted = 0;
  let eventsRejected = 0;
  let eventsDuplicate = 0;

  let status = "success";
  /** @type {Date|null} */
  let watermark = null;
  /** @type {SensorEvent[]} */
  const lateEvents = [];
  let maxLateness = 0;

  try {
    const raw = await readFile(inputPath, "utf-8");
    const lines = raw.split(/\r?\n/).filter((line) => line.trim().length > 0);

    for (const line of lines) {
      eventsSeen += 1;

      // Parse JSON
      let event;
      try {
        event = JSON.parse(line);
      } catch {
        const reason = "malformed JSON";
        eventsRejected += 1;
        rejectionReasons[reason] = (rejectionReasons[reason] || 0) + 1;
        deadLetterEvents.push({ event: line, status: "rejected", reason });
        console.warn(`[${pipelineName}] Rejected line ${eventsSeen}: ${reason}`);
        continue;
      }

      // Validate
      let result;
      try {
        result = validateEvent(event);
      } catch (err) {
        const reason = `validation error: ${err.message}`;
        eventsRejected += 1;
        rejectionReasons[reason] = (rejectionReasons[reason] || 0) + 1;
        deadLetterEvents.push({ event, status: "rejected", reason });
        console.warn(`[${pipelineName}] Rejected line ${eventsSeen}: ${reason}`);
        continue;
      }

      if (result.status === "rejected") {
        eventsRejected += 1;
        rejectionReasons[result.reason] = (rejectionReasons[result.reason] || 0) + 1;
        deadLetterEvents.push({ event: result.event, status: "rejected", reason: result.reason });
        console.warn(`[${pipelineName}] Rejected line ${eventsSeen}: ${result.reason}`);
        continue;
      }

      // Deduplicate
      const key = deduplicateKey(event);
      if (seenKeys.has(key)) {
        const reason = "duplicate event";
        eventsDuplicate += 1;
        rejectionReasons[reason] = (rejectionReasons[reason] || 0) + 1;
        deadLetterEvents.push({ event, status: "duplicate", reason });
        console.warn(`[${pipelineName}] Duplicate line ${eventsSeen}: ${key}`);
        continue;
      }
      seenKeys.add(key);

      // Accept
      eventsAccepted += 1;
      acceptedEvents.push(event);

      // Lateness check
      const eventTime = parseEventTime(event.timestamp);
      const { is_late: isLate, lateness_seconds: lateness } = classifyLateness(
        eventTime, watermark, latenessThresholdSeconds,
      );
      if (isLate) {
        lateEvents.push(event);
        if (lateness > maxLateness) maxLateness = lateness;
      }
      // Advance watermark
      if (watermark === null || eventTime > watermark) {
        watermark = eventTime;
      }
    }
  } catch (err) {
    status = "failed";
    console.error(`[${pipelineName}] Fatal error: ${err.message}`);
  }

  // Write accepted events
  const acceptedPath = join(outputDir, "accepted.jsonl");
  const acceptedContent = acceptedEvents.map((e) => JSON.stringify(e)).join("\n");
  await writeFile(acceptedPath, acceptedContent.length > 0 ? acceptedContent + "\n" : "", "utf-8");

  // Write dead-letter events
  const deadLetterPath = join(outputDir, "dead_letter.jsonl");
  const deadLetterContent = deadLetterEvents.map((e) => JSON.stringify(e)).join("\n");
  await writeFile(deadLetterPath, deadLetterContent.length > 0 ? deadLetterContent + "\n" : "", "utf-8");

  // Write late events
  const lateEventsPath = join(outputDir, "late_events.jsonl");
  const lateEventsContent = lateEvents.map((e) => JSON.stringify(e)).join("\n");
  await writeFile(lateEventsPath, lateEventsContent.length > 0 ? lateEventsContent + "\n" : "", "utf-8");

  // Compute aggregates
  const aggregates = computeAggregates(acceptedEvents);

  const endTime = performance.now();
  const durationSeconds = Math.round(((endTime - startTime) / 1000) * 1000) / 1000;

  /** @type {PipelineSummary} */
  const summary = {
    pipeline_name: pipelineName,
    run_at: runAt,
    duration_seconds: durationSeconds,
    status,
    events_seen: eventsSeen,
    events_accepted: eventsAccepted,
    events_rejected: eventsRejected,
    events_duplicate: eventsDuplicate,
    dead_letter_count: eventsRejected + eventsDuplicate,
    events_late: lateEvents.length,
    max_lateness_seconds: Math.round(maxLateness * 100) / 100,
    watermark: watermark ? watermark.toISOString() : "",
    lateness_threshold_seconds: latenessThresholdSeconds,
    aggregates,
    rejection_reasons: rejectionReasons,
  };

  // Write summary
  const summaryPath = join(outputDir, "summary.json");
  await writeFile(summaryPath, JSON.stringify(summary, null, 2) + "\n", "utf-8");

  console.info(
    `[${pipelineName}] Complete: ${eventsAccepted} accepted (${lateEvents.length} late), ` +
    `${eventsRejected} rejected, ${eventsDuplicate} duplicate ` +
    `(${durationSeconds}s)`,
  );

  return summary;
}
